File: benchmarks/cross-platform/overhead_test.py
#!/usr/bin/env python3
"""
Test to measure Python/NumPy call overhead vs actual computation.
"""
import time
import numpy as np

# Create test data
data = np.random.randn(10000, 100)

# Measure individual operation overhead
def measure_overhead():
    results = {}

    # 1. Just the function call overhead (tiny array)
    tiny = np.array([1.0])

    start = time.perf_counter()
    for _ in range(100000):
        np.mean(tiny)
    end = time.perf_counter()
    results['mean(tiny) per call'] = (end - start) / 100000 * 1e6  # microseconds

    # 2. Mean on large array
    start = time.perf_counter()
    for _ in range(1000):
        np.mean(data)
    end = time.perf_counter()
    results['mean(10k x 100) per call'] = (end - start) / 1000 * 1e6

    # 3. Z-score as separate operations
    start = time.perf_counter()
    for _ in range(100):
        m = np.mean(data, axis=0)
        s = np.std(data, axis=0)
        result = (data - m) / s
    end = time.perf_counter()
    results['zscore separate ops'] = (end - start) / 100 * 1000  # ms

    # 4. Count the operations
    results['zscore op count'] = '4 ops: mean, std, subtract, divide'

    return results

results = measure_overhead()

print("Python/NumPy Call Overhead Analysis")
print("=" * 50)
for k, v in results.items():
    if isinstance(v, float):
        print(f"{k}: {v:.2f} µs" if v < 1000 else f"{k}: {v/1000:.2f} ms")
    else:
        print(f"{k}: {v}")

print("\nKey insight:")
print(f"  - Tiny array mean: ~{results['mean(tiny) per call']:.1f} µs (mostly overhead)")
print(f"  - Large array mean: ~{results['mean(10k x 100) per call']:.0f} µs (computation dominates)")
print(f"  - Overhead ratio: ~{results['mean(10k x 100) per call'] / results['mean(tiny) per call']:.1f}x more work, but overhead is amortized")

File: benchmarks/cross-platform/overhead_test.ts
#!/usr/bin/env npx tsx
/**
 * Test to measure N-API call overhead vs actual computation.
 */
import { array, mean, std, subtract, divide, zeros } from '../../src/index.js';

// Helper to create random matrix
function randomMatrix(rows: number, cols: number): number[][] {
  const data: number[][] = [];
  for (let i = 0; i < rows; i++) {
    const row: number[] = [];
    for (let j = 0; j < cols; j++) {
      row.push(Math.random() * 2 - 1);
    }
    data.push(row);
  }
  return data;
}

// Create test data
const data = array(randomMatrix(10000, 100));

function measureOverhead() {
  const results: Record<string, number | string> = {};

  // 1. Just the function call overhead (tiny array)
  const tiny = array([1.0]);

  let start = performance.now();
  for (let i = 0; i < 100000; i++) {
    mean(tiny);
  }
  let end = performance.now();
  results['mean(tiny) per call'] = ((end - start) / 100000) * 1000; // microseconds

  // 2. Mean on large array
  start = performance.now();
  for (let i = 0; i < 1000; i++) {
    mean(data);
  }
  end = performance.now();
  results['mean(10k x 100) per call'] = ((end - start) / 1000) * 1000;

  // 3. Z-score as separate operations
  start = performance.now();
  for (let i = 0; i < 100; i++) {
    const m = mean(data, 0);
    const s = std(data, 0);
    const centered = subtract(data, m);
    const result = divide(centered, s);
  }
  end = performance.now();
  results['zscore separate ops'] = (end - start) / 100; // ms

  // 4. Count the operations
  results['zscore op count'] = '4 ops: mean, std, subtract, divide';

  return results;
}

const results = measureOverhead();

console.log('Node.js/numpy-node Call Overhead Analysis');
console.log('='.repeat(50));
for (const [k, v] of Object.entries(results)) {
  if (typeof v === 'number') {
    console.log(`${k}: ${v < 1000 ? v.toFixed(2) + ' µs' : (v / 1000).toFixed(2) + ' ms'}`);
  } else {
    console.log(`${k}: ${v}`);
  }
}

const tinyOverhead = results['mean(tiny) per call'] as number;
const largeTime = results['mean(10k x 100) per call'] as number;

console.log('\nKey insight:');
console.log(`  - Tiny array mean: ~${tinyOverhead.toFixed(1)} µs (mostly N-API overhead)`);
console.log(`  - Large array mean: ~${largeTime.toFixed(0)} µs`);
console.log(`  - Overhead ratio: ${(largeTime / tinyOverhead).toFixed(1)}x more work`);
